### PR TITLE
[SPARK-2489][SQL] Support Parquet's optional fixed_len_byte_array

### PR DESCRIPTION
**Pull request by aws-awinstan**
## What changes were proposed in this pull request?
This PR adds support for reading Parquet FIXED_LENGTH_BYTE_ARRAYs as a Binary column if no OriginalType is specified. Parquet-avro writes the Avro fixed type as a Parquet FIXED_LENGTH_BYTE_ARRAY type. Currently when trying to load Parquet files with a column of this type with Spark SQL it throws an exception similar to the following:

```
Caused by: org.apache.spark.sql.AnalysisException: Illegal Parquet type: FIXED_LEN_BYTE_ARRAY;
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.illegalType$1(ParquetSchemaConverter.scala:108)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convertPrimitiveField(ParquetSchemaConverter.scala:177)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convertField(ParquetSchemaConverter.scala:90)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter$$anonfun$2.apply(ParquetSchemaConverter.scala:72)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter$$anonfun$2.apply(ParquetSchemaConverter.scala:66)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.org$apache$spark$sql$execution$datasources$parquet$ParquetToSparkSchemaConverter$$convert(ParquetSchemaConverter.scala:66)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.convert(ParquetSchemaConverter.scala:63)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anonfun$readSchemaFromFooter$2.apply(ParquetFileFormat.scala:642)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anonfun$readSchemaFromFooter$2.apply(ParquetFileFormat.scala:642)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$.readSchemaFromFooter(ParquetFileFormat.scala:642)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anonfun$9.apply(ParquetFileFormat.scala:599)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anonfun$9.apply(ParquetFileFormat.scala:581)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:800)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:800)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

After this change Spark SQL is able to correctly load the Parquet files. There was a PR to fix this 3 years ago (https://github.com/apache/spark/pull/1737) however it was ultimately rejected as the committer went down the path of adding a new SQL Type specifically for FIXED_LENGTH_BYTE_ARRAYs and the maintainers believed this was too intrusive of a change. This PR simply defaults to Binary if no OriginalType is specified. A few updates were required to the VectorizedColumnReader to support Binary FIXED_LENGTH_BYTE_ARRAYs.

Note: All the changes to the gen-java/* files were generated by avro-tools-1.8.1 and the mostly documentation updates look to come from changes in the template avro-tools uses.

## How was this patch tested?

I added a fixed attribute to the AvroPrimitives and AvroOptionalPrimitives record types which are used by the ParquetAvroCompatibilitySuite. These values were populated by taking the same value as other type ("val_$i"), padding it to 8 bytes (the chosen fixed length), and storing it as the fixed type. I verified that before my fix the "required primitives" and "optional primitives" failed with the same exception we're seeing in our clusters. After my change the tests succeed with the expected results.

### CODE REVIEW

File: sql/core/src/test/gen-java/org/apache/spark/sql/execution/datasources/parquet/test/avro/AvroArrayOfArray.java [hunk 1]
```diff
@@ -1,39 +1,45 @@
...
+  private static final long serialVersionUID = -5085901832173924180L;
   public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"AvroArrayOfArray\",\"namespace\":\"org.apache.spark.sql.execution.datasources.parquet.test.avro\",\"fields\":[{\"name\":\"int_arrays_column\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"array\",\"items\":\"int\"}}}]}");
   public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
   @Deprecated public java.util.List<java.util.List<java.lang.Integer>> int_arrays_column;
 
   /**
    * Default constructor.  Note that this does not initialize fields
    * to their default values from the schema.  If that is desired then
-   * one should use <code>newBuilder()</code>. 
+   * one should use <code>newBuilder()</code>.
```
Comment (HyukjinKwon) [inline sql/core/src/test/gen-java/org/apache/spark/sql/execution/datasources/parquet/test/avro/AvroArrayOfArray.java]: @aws-awinstan Can we remove unrelated changes? It looks hard to follow the changes.

Comment (aws-awinstan) [reply sql/core/src/test/gen-java/org/apache/spark/sql/execution/datasources/parquet/test/avro/AvroArrayOfArray.java]: @HyukjinKwon Done! I had debated leaving these changes in when submitting the PR but decided to leave the newly generated files since they had some minor improvements (e.g. to the JavaDocs). Ideally these Avro files would be generated during the build process rather than checked in but that's a separate issue.

**Comment by aws-awinstan** [comment]
Just pinging on this for a review. Let me know if there are any questions or concerns.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @liancheng and @marmbrus too.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by aws-awinstan** [comment]
@HyukjinKwon @liancheng @marmbrus - Any comments on this PR? Can we get this merged?

**Comment by spetsnaz84** [comment]
Hope we can see this in the next release of Spark...

**Comment by aws-awinstan** [comment]
Pinging on this once again. It's been almost a year since this PR was opened.

**Comment by praetp** [comment]
Can this be merged in ?

**Comment by HyukjinKwon** [comment]
cc @rdblue for review since it;s a Parquet one.

**Comment by ozars** [comment]
Pinging this. Could you please have a look at this PR?

**Comment by rdblue** [comment]
+1

This looks good to me.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 2]
```diff
@@ -509,6 +509,14 @@ private void readFixedLenByteArrayBatch(
           column.putNull(rowId + i);
         }
       }
+    } else if (column.dataType() == DataTypes.BinaryType) {
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: Looks like this change doesn't have associated test. Could you add one?

**Comment by viirya** [comment]
I am not sure about the comment https://github.com/apache/spark/pull/1737#issuecomment-53976708 said before. It seems a concern about using BinaryType for fixed_len_byte_array. What do you think about the argument?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetAvroCompatibilitySuite.scala [hunk 3]
```diff
@@ -62,12 +70,12 @@ class ParquetAvroCompatibilitySuite extends ParquetCompatibilityTest with Shared
...
               .setBinaryColumn(ByteBuffer.wrap(s"val_$i".getBytes(StandardCharsets.UTF_8)))
               .setStringColumn(s"val_$i")
+              .setFixedColumn(new FixedType(generateFixedLengthByteArray(i)))
               .build())
         }
       }
 
       logParquetSchema(path)
-
       checkAnswer(spark.read.parquet(path), (0 until 10).map { i =>
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetAvroCompatibilitySuite.scala]: I think this only tests against `spark.sql.parquet.enableVectorizedReader` is true (default), should we test non-vectorized reader too?

**Comment by ozars** [comment]
@rdblue @viirya  Thanks for checking it out.

> I am not sure about the comment [#1737 (comment)](https://github.com/apache/spark/pull/1737#issuecomment-53976708) said before. It seems a concern about using BinaryType for fixed_len_byte_array. What do you think about the argument?

I bisected for `convertFromAttributes` and figured out the code referred by that message (or at least some portion of it) was refactored by 02149ff08eed3745086589a047adbce9a580389f in 2015. I'm not very familiar with the internal machinery of spark. Could someone familiar with the codebase confirm if this is still a concern? 

cc @liancheng

**Comment by rdblue** [comment]
I think it is fine to use BinaryType to read fixed_len_byte_array. That's the best mapping, even though the type is technically wider because it includes shorter and longer binary sequences. This only allows reading that data, not writing values with a fixed length. All writes would use BinaryType regardless.

**Comment by viirya** [comment]
Thanks for clarifying. Although that comment concerned about reading problem, looks like `convertFromAttributes` was for writing.  This PR works on reading fixed_len_byte_array as Spark BinaryType. I think it should be fine. cc @cloud-fan too

**Comment by cloud-fan** [comment]
The idea is fine. Spark reads Hive varchar as string type, although it's not an accurate mapping.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetAvroCompatibilitySuite.scala [hunk 4]
```diff
@@ -47,6 +47,14 @@ class ParquetAvroCompatibilitySuite extends ParquetCompatibilityTest with Shared
     try f(writer) finally writer.close()
   }
 
+  private def generateFixedLengthByteArray(i : Int): Array[Byte] = {
+    val fixedLengthByteArray = Array[Byte](0, 0, 0, 0, 0, 0, 0, 0)
+    val fixedLengthByteArrayComponent = s"val_$i".getBytes(StandardCharsets.UTF_8)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetAvroCompatibilitySuite.scala]: nit `s` seems not needed

**Comment by HyukjinKwon** [comment]
Yea, let's have a parquet dedicated test (I guess) as already pointed out.

**Review by HyukjinKwon — COMMENTED** [review]
I am okay with supporting this - don't particularly mind.

But I would like to note:

https://github.com/apache/spark/blob/27ac6af30cde9cd596929ab8fd076c9b41287384/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L138-L140

We don't support unsigned types although other types like long can contain (see https://github.com/apache/spark/pull/9646 - wow it's 4 years ago). cc @liancheng

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: We can merge it into the previous if:
```
else if (DecimalType.isByteArrayDecimalType(column.dataType()) || column.dataType() == DataTypes.BinaryType)
```

**Comment by cloud-fan** [comment]
The added tests only test parquet-avro compatibility. We should have a test for parquet.

**Comment by cloud-fan** [comment]
Is anyone interested in this PR and want to take over?

**Comment by jonbelanger-ns** [comment]
If it helps, I have a fairly complex parquet file with a few nest fields as FIXED_LEN_BYTE_ARRAY, so this bug is a show stopper for spark on this dataset.

I tried to fix by cloning this repo with the PR (https://github.com/aws-awinstan/spark.git) to local machine and compiling.

I did the same for the master repo for spark which worked fine on a with a few of the columns (to test without parsing the FIXED_LEN_BYTE_ARRAY columns).

However, the aws-awinstan repo fails with on the same test columns:

[Stage 0:>                                                          (0 + 1) / 1]20/01/17 12:37:13 WARN TaskSetManager: Lost task 0.0 in stage 0.0 (TID 0, 192.168.42.107, executor 0): java.io.StreamCorruptedException: invalid stream header: 0000000F
	at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:866)
	at java.io.ObjectInputStream.<init>(ObjectInputStream.java:358)
	at org.apache.spark.serializer.JavaDeserializationStream$$anon$1.<init>(JavaSerializer.scala:63)
	at org.apache.spark.serializer.JavaDeserializationStream.<init>(JavaSerializer.scala:63)
	at org.apache.spark.serializer.JavaSerializerInstance.deserializeStream(JavaSerializer.scala:126)
	at org.apache.spark.serializer.JavaSerializerInstance.deserialize(JavaSerializer.scala:113)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:313)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

I'm using in the following in my client environment, with the HDFS and Spark remote in VM and standalone with a single worker.

$ pip freeze | grep spark
pyspark==2.4.4
spark==0.2.1

I'm surprised this bug was allowed to languish for as long as it has, it's not possible for us to serialize the upstream data and need this feature or have to move on...

**Comment by nicolaslrveiga** [comment]
Can we reopen this issue or proceed with the work in https://github.com/apache/spark/pull/35902?
